### PR TITLE
Disallow setter-only EqualityContract property

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/48723

### CODE REVIEW

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs [hunk 1]
```diff
@@ -56,6 +56,13 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
 
             try
             {
+                if (_equalityContract.GetMethod is null)
+                {
+                    // There the equality contract isn't usable, an error was reported elsewhere
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:61]: >There the equality contract [](start = 23, length = 27)

"The equality contract ..."?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordGetHashCode.cs [hunk 2]
```diff
@@ -47,6 +47,13 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
 
             try
             {
+                if (_equalityContract.GetMethod is null)
+                {
+                    // There the equality contract isn't usable, an error was reported elsewhere
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordGetHashCode.cs:52]: >There the equality contract [](start = 23, length = 27)

"The equality contract ..."?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 3]
```diff
@@ -1921,6 +1921,7 @@ internal enum ErrorCode
         #endregion diagnostics introduced for C# 9.0
 
         ERR_UnexpectedVarianceStaticMember = 9100,
+        ERR_EqualityContractRequiresGetter = 9101,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1924]: >ERR_EqualityContractRequiresGetter = 9101, [](start = 8, length = 42)

It looks like this branch is significantly behind. I think in master the last used error number is 8905

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs [hunk 4]
```diff
@@ -56,6 +56,13 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
 
             try
             {
+                if (_equalityContract.GetMethod is null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:59]: >if (_equalityContract.GetMethod is null) [](start = 16, length = 40)

Should we limit this check to the code path that actually uses the property?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordGetHashCode.cs [hunk 5]
```diff
@@ -47,6 +47,13 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
 
             try
             {
+                if (_equalityContract.GetMethod is null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordGetHashCode.cs:50]: >if (_equalityContract.GetMethod is null) [](start = 16, length = 40)

Should we limit this check to the code path that actually uses the property?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 6]
```diff
@@ -18548,6 +18548,25 @@ record A
...
         }
 
+        [Fact]
+        [WorkItem(48723, "https://github.com/dotnet/roslyn/issues/48723")]
+        public void EqualityContract_24_SetterOnlyProperty()
+        {
+            var src = @"
+record R
+{
+    protected virtual System.Type EqualityContract { set { } }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18558]: >set { } [](start = 53, length = 7)

Consider adding a test with a property that has both accessors, unless we already have one. I assume that should be a success scenario? Or should we disallow property like that?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18558]: Also, consider adding test(s) when a record class has only setter or both accessors in metadata and we try to derive another record from it with explicit/implicit definition of EqualityContract. 

---
In reply to: [508933817](https://github.com/dotnet/roslyn/pull/48798#discussion_r508933817) [](ancestors = 508933817)

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18558]: > Or should we disallow property like that [with both accessors]?

I think it's okay to allow, as it meets our requirements and doesn't get in the way of what we need to do. 

---
In reply to: [508941602](https://github.com/dotnet/roslyn/pull/48798#discussion_r508941602) [](ancestors = 508941602,508933817)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 7]
```diff
@@ -18548,6 +18548,282 @@ record A
...
+}
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyEmitDiagnostics(
+                // (4,35): error CS8906: Record equality contract property 'R.EqualityContract' must have a get accessor.
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18564]: equality contract seems unnecessarily repetitive in the message.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 8]
```diff
@@ -18548,6 +18548,282 @@ record A
...
+            comp2.VerifyEmitDiagnostics(
+                // (4,56): error CS0545: 'R.EqualityContract.get': cannot override because 'Base.EqualityContract' does not have an overridable get accessor
+                //     protected override System.Type EqualityContract => typeof(R);
+                Diagnostic(ErrorCode.ERR_NoGetToOverride, "typeof(R)").WithArguments("R.EqualityContract.get", "Base.EqualityContract").WithLocation(4, 56)
+                );
+        }
+
+        [Fact]
+        [WorkItem(48723, "https://github.com/dotnet/roslyn/issues/48723")]
+        public void EqualityContract_27_GetterAndSetterOnlyProperty_InMetadata()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18720]: >Only [](start = 55, length = 4)

Is the "Only" part in the name kept to convey something important?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 9]
```diff
@@ -18548,6 +18548,282 @@ record A
...
+    }
+
+    .method family hidebysig specialname newslot virtual instance void set_EqualityContract ( class [mscorlib]System.Type 'value' ) cil managed
+    {
+        IL_0000: ldnull
+        IL_0001: throw
+    }
+
+    // Property has a getter and setter
+    .property instance class [mscorlib]System.Type EqualityContract()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18801]: >.property instance class [mscorlib]System.Type EqualityContract() [](start = 4, length = 65)

It looks like we decided to allow a declaration like that in source. Is there a particular reason to use IL to define this property for the test?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 10]
```diff
@@ -18548,6 +18548,282 @@ record A
...
+    }
+}
+";
+
+            var src = @"
+record R : Base;
+";
+
+            var comp = CreateCompilationWithIL(src, il);
+            comp.VerifyEmitDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18814]: > comp.VerifyEmitDiagnostics(); [](start = 11, length = 30)

I think that for scenarios tested in this test we would want to test runtime behavior in order to verify that overriding is indeed happening and has expected effect on equality behavior.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3) with some comments for tests.

**Pull request merged**